lib/getenv.c

```diff
@@ -53,15 +56,20 @@ static char *GetEnv(const char *variable)
 
     /* It's possible for rc to be 0 if the variable was found but empty.
        Since getenv doesn't make that distinction we ignore it as well. */
-    rc = GetEnvironmentVariableA(variable, buf, bufsize);
+    rc = GetEnvironmentVariable(tchar_envvar, buf, bufsize);
     if(!rc || rc == bufsize || rc > max) {
       free(buf);
+      curlx_unicodefree(tchar_envvar);
       return NULL;
     }
 
     /* if rc < bufsize then rc is bytes written not including null */
-    if(rc < bufsize)
-      return buf;
+    if(rc < bufsize) {
+      char *path = curlx_convert_tchar_to_UTF8(buf);
+      curlx_unicodefree(buf);
+      curlx_unicodefree(tchar_envvar);
+      return path;
```

Comment:
You're assuming path will be heap allocated and currently that is [not guaranteed](https://github.com/curl/curl/blob/curl-7_71_1/lib/curl_multibyte.h#L49-L52) unless UNICODE and WIN32. If ANSI it returns the passed in pointer, in other words it could be that `path == buf` so in that case you wouldn't free buf. Also IIRC this is the internal function used for the public api curl_getenv. In this case I don't know that we'd want to convert the environment variable value to UTF-8 encoding. That would be a change in behavior. curl_getenv has delivered ansi encoding and probably should continue to do so.

I really don't like how we have those conversion macros sometimes returning a copy on the heap and sometimes returning just the original pointer, depending on how libcurl was built. I assume it was done for performance but it can lead to mistakes like this, because it's easy to forget that it may or may not return a copy.

Best resolution for this in my opinion is preempt it with a separate fix to always return a copy on the heap. So in other words change the macro to do a strdup for ansi builds.

(There may be other cases in your changes where this is also an issue but skimming it just now I don't see any)

Reply:
I don't see a problem here: with ANSI, we'll get `buf` like before and it won't get freed here. With Unicode, we'll get a copy and the original `wchar_t*` will get freed.

Yeah, always returning a copy would make many things easier and remove the need for `curlx_unicodefree`.

You're right about the behavior change in the public API. I'll revert this one.

Reply:
> I don't see a problem here: with ANSI, we'll get `buf` like before and it won't get freed here. With Unicode, we'll get a copy and the original `wchar_t*` will get freed.

Oops you're right I misread it. I thought that the free macro actually freed it, but if ANSI it doesn't because it doesn't make the copy.